homeassistant/components/ultrasync/sensor.py

```diff
@@ -0,0 +1,176 @@
+                new_sensors.append(sensors[sensor_id])
+                _LOGGER.debug(
+                    "Detected %s.Area%dState", entry.data[CONF_NAME], bank_no + 1
+                )
+
+            # Update our meta information
+            for key, value in meta.items():
+                sensors[sensor_id][key] = value
+
+        for meta in zones:
+            bank_no = meta["bank"]
+            sensor_id = "zone{:0>2}_state".format(bank_no + 1)
+            detected_sensors.add(sensor_id)
+            if sensor_id not in sensors:
+                # hash our entry
+                sensors[sensor_id] = UltraSyncSensor(
+                    coordinator,
+                    entry.entry_id,
+                    entry.data[CONF_NAME],
+                    sensor_id,
+                    # Friendly Name
+                    "Zone{}State".format(bank_no + 1),
+                )
+
+                # Add our new zone sensor
+                new_sensors.append(sensors[sensor_id])
+                _LOGGER.debug(
+                    "Detected %s.Zone%dState", entry.data[CONF_NAME], bank_no + 1
+                )
+
+            # Update our meta information
+            for key, value in meta.items():
+                sensors[sensor_id][key] = value
+
+        if new_sensors:
+            # Add our newly detected sensors
+            async_add_entities(new_sensors)
+
+        for sensor_id in set(sensors.keys()).difference(detected_sensors):
+            # Tidy up sensors leaving our listing
+            hass.async_create_task(sensors[sensor_id].async_remove())
+            del sensors[sensor_id]
+
+    # register our callback which will be called the second we make a
+    # connection to our panel
+    hass.data[DOMAIN][entry.entry_id][DATA_UNDO_UPDATE_LISTENER].append(
+        async_dispatcher_connect(hass, SENSOR_UPDATE_LISTENER, _auto_manage_sensors)
+    )
+
+
+class UltraSyncSensor(UltraSyncEntity):
+    """Representation of a UltraSync sensor."""
+
+    def __init__(
+        self,
+        coordinator: UltraSyncDataUpdateCoordinator,
+        entry_id: str,
+        entry_name: str,
+        sensor_type: str,
+        sensor_name: str,
+    ):
+        """Initialize a new UltraSync sensor."""
+
+        self._sensor_type = sensor_type
+        self._unique_id = f"{entry_id}_{sensor_type}"
+
+        # Initialize our Attributes
+        self.__attributes = {}
+
+        super().__init__(
+            coordinator=coordinator,
+            entry_id=entry_id,
+            name=f"{entry_name} {sensor_name}",
+        )
+
+    def __setitem__(self, key, value):
+        """Set our sensor attributes."""
+        self.__attributes[key] = value
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID of the sensor."""
+        return self._unique_id
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        return self.__attributes
```

Comment:
What are the possible attributes? You need to specify a list/dict of possible attributes, so it is clear what will be added. This cannot be dynamic.

Reply:
The problem is that they truly are dynamic.  The ultrasync is dynamically configured from the panel or it's internal website.  I don't knjow what is enabled and disabled until i first connect to the panel (using the pin/password provided by the configuration).

Once i can connect, i can then poll for all of the zones and sensors.  Different sensors have different attributes associated with them (such as a sensor like a smoke detector, motion sensors, and door sensors, etc).  So I'm able to poll this and build the sensor list dynamically.  It actually works really well; I've had lots of positive feedback using HACS in the time being since [here](https://github.com/caronc/ha-ultrasync).